clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1146,6 +1204,53 @@ void KernelObjVisitor::visitNthArrayElement(const CXXRecordDecl *Owner,
             .Handler...);
 }
 
+template <typename ParentTy, typename... HandlerTys>
+void KernelObjVisitor::visitRecord(const CXXRecordDecl *Owner, ParentTy &Parent,
+                                   const CXXRecordDecl *Wrapper,
+                                   QualType RecordTy,
+                                   HandlerTys &... Handlers) {
+  if (RecordTy->getAsRecordDecl()->hasAttr<SYCLRequiresDecompositionAttr>()) {
+    visitComplexRecord(Owner, Parent, Wrapper, RecordTy, Handlers...);
+  } else {
+    if (!AllTrue<HandlerTys::VisitInsideSimpleContainers...>::Value)
+      visitSimpleRecord(
+          Owner, Parent, Wrapper, RecordTy,
+          HandlerFilter<!HandlerTys::VisitInsideSimpleContainers, HandlerTys>(
+              Handlers)
+              .Handler...);
+
+    if (AnyTrue<HandlerTys::VisitInsideSimpleContainers...>::Value)
+      visitComplexRecord(
```

Comment:
So VisitSimpleRecord is called if any (one or more) handler has opted out and VisitComplexRecord is called if any has 'opted in'? 

Is this required because the visitor runs through twice? First with checkers/attribute marking and second time with declcreator/bodycreator/intheader (so all opted out and VisitSimpleRecord is called)? I don't fully understand the need for this logic. Could you give and example for each case?

Reply:
The 'if' statements here are a little bit of an optimization, they aren't technically necessary.

VisitSimpleRecord is called ONLY on handlers that have  VisitInsideSimpleContainers=false (see the HandlerFilter).

VisitComplexRecord is called ONLY on handlers that have VisitInsideSimpleContainers=true (via the Handler Filter).

The 'if' simply makes sure that we ONLY call this when the HandlerFilter doesn't end up giving us an empty set.

Reply:
BTW, comment added in the incoming patch.